### PR TITLE
Fog of war

### PR DESCRIPTION
**Pull request by jorisBarkema**
added fog of war

### CODE REVIEW

**Review by jensj12 — CHANGES_REQUESTED** [review]
Since the fogOfWar bool should be able to override the visibility settings of the tile, I prefer a new bool for the tile. That also makes it possible to toggle fog of war in-game. Override the draw function, check that bool and call the base function. The draw function will be overridden anyways because some tiles need a secondary sprite.
Anyways, I love it.

**Comment by jorisBarkema** [comment]
Tiles now has a fogOfWar and a visited bool, the visited bool is necessary for togglability. Tiles also has a toggleFogOfWar method, to toggle the fog of war just call the toggleFogOfWar method for each tile. Did not add a method for that anywhere yet as the toggling will go through the UI and I don't know exactly how that's going to interact, so where that method should be.

editL some bugs right now, working on it

**Comment by jorisBarkema** [comment]
so the bug was...I set visible to false when the tile was visited.....

**Comment by jorisBarkema** [comment]
I think this is what you were talking about.

File: MeesGame/Gameplay/GameObjects/Tile.cs [hunk 1]
```diff
@@ -16,10 +17,24 @@ abstract class Tile : SpriteGameObject
     {
         protected TileType tileType;
         protected Point location = Point.Zero;
+        protected bool revealed = false;
 
-        protected Tile(string assetName = "", TileType tileType = TileType.Floor, int layer = 0, string id = "") : base(assetName, layer, id)
+        protected Tile(bool fogOfWar = true, string assetName = "", TileType tileType = TileType.Floor, int layer = 0, string id = "") : base(assetName, layer, id)
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/Tile.cs]: Don't add this parameter to the constructor (same for all tile subclasses).

**Review by jensj12 — CHANGES_REQUESTED** [review]

**Comment by jorisBarkema** [comment]
forgot about those, fixed now

File: MeesGame/Gameplay/GameObjects/TileField.cs [hunk 2]
```diff
@@ -4,9 +4,23 @@ namespace MeesGame
 {
     class TileField : GameObjectGrid
     {
-        public TileField(int rows, int columns, int layer = 0, string id = "") : base(rows, columns, layer, id)
+        protected bool fogOfWar;
+
+        public TileField(int rows, int columns, int layer = 0, bool fogOfWar = true, string id = "") : base(rows, columns, layer, id)
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/TileField.cs]: Place fogOfWar before layer

File: MeesGame/Gameplay/GameObjects/Tile.cs [hunk 3]
```diff
@@ -39,6 +54,14 @@ public Point Location
             set { location = value; }
         }
 
+        public override void Draw(GameTime gameTime, SpriteBatch spriteBatch)
+        {
+            if ((!(this.Parent as TileField).FogOfWar || this.Revealed) && sprite != null)
+            {
+                sprite.Draw(spriteBatch, this.GlobalPosition, origin);
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/Tile.cs]: Call base.Draw and don't check for sprite!=null

File: MeesGame/Gameplay/GameObjects/DoorTile.cs [hunk 4]
```diff
@@ -7,7 +7,7 @@ class DoorTile : WallTile
         protected bool doorIsOpen = false;
 
         //TODO: Pair a door with a specific key.
-        public DoorTile(int layer = 0, string id = "") : base("horizontalDoorOverlay", TileType.Door, layer, id)
+        public DoorTile(bool fogOfWar = true, int layer = 0, string id = "") : base(fogOfWar, "horizontalDoorOverlay", TileType.Door, layer, id)
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/DoorTile.cs]: Remove fogOfWar

File: MeesGame/Gameplay/GameObjects/DoorTile.cs [hunk 5]
```diff
@@ -63,7 +63,7 @@ public override void UpdateGraphicsToMatchSurroundings()
     class KeyTile : FloorTile
     {
 
-        public KeyTile(int layer = 0, string id = "") : base("keyOverlay", TileType.Key, layer, id)
+        public KeyTile(bool fogOfWar = true, int layer = 0, string id = "") : base(fogOfWar, "keyOverlay", TileType.Key, layer, id)
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/DoorTile.cs]: Remove fogOfWar

**Review by jensj12 — CHANGES_REQUESTED** [review]

**Review by jensj12 — APPROVED** [review]

**Pull request merged**